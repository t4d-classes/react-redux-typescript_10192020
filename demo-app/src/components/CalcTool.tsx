import React, { Component, ChangeEvent } from 'react';

export type CalcToolProps = {
  result: number;
  onAdd: (num: number) => void;
  onSubtract: (num: number) => void;
};

type CalcToolState = {};

export class CalcTool extends Component<CalcToolProps, CalcToolState> {
  state = {
    numInput: 0,
  };

  change = (e: ChangeEvent<HTMLInputElement>) => {
    this.setState({
      numInput: Number(e.target.value),
    });
  };

  render() {
    return (
      <form>
        <div>Result: {this.props.result}</div>
        <div>
          Num Input:
          <input
            type="number"
            value={this.state.numInput}
            onChange={this.change}
          />
        </div>
        <fieldset>
          <button
            type="button"
            onClick={() => this.props.onAdd(this.state.numInput)}>
            +
          </button>
          <button
            type="button"
            onClick={() => this.props.onSubtract(this.state.numInput)}>
            -
          </button>
        </fieldset>
      </form>
    );
  }
}
