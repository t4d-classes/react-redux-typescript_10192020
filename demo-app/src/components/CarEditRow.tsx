import React, { Component, ChangeEvent } from 'react';

import { Car } from '../models/cars';

export type CarEditRowProps = {
  car: Car;
  onSaveCar: (car: Car) => void;
  onCancelCar: () => void;
};

export type CarEditRowState = {
  make: string;
  model: string;
  year: number;
  color: string;
  price: number;
};

export class CarEditRow extends Component<CarEditRowProps, CarEditRowState> {
  state = {
    make: this.props.car.make,
    model: this.props.car.model,
    year: this.props.car.year,
    color: this.props.car.color,
    price: this.props.car.price,
  };

  change = (e: ChangeEvent<HTMLInputElement>) => {
    this.setState({
      [e.target.name]:
        e.target.type === 'number' ? Number(e.target.value) : e.target.value,
    } as Pick<CarEditRowState, keyof CarEditRowState>);
  };

  saveCar = () => {
    this.props.onSaveCar({ ...this.state, id: this.props.car.id });
  };

  render() {
    return (
      <tr>
        <td>{this.props.car.id}</td>
        <td>
          <input
            type="text"
            id="make-input"
            name="make"
            value={this.state.make}
            onChange={this.change}
          />
        </td>
        <td>
          <input
            type="text"
            id="model-input"
            name="model"
            value={this.state.model}
            onChange={this.change}
          />
        </td>
        <td>
          <input
            type="number"
            id="year-input"
            name="year"
            value={this.state.year}
            onChange={this.change}
          />
        </td>
        <td>
          <input
            type="text"
            id="color-input"
            name="color"
            value={this.state.color}
            onChange={this.change}
          />
        </td>
        <td>
          <input
            type="number"
            id="price-input"
            name="price"
            value={this.state.price}
            onChange={this.change}
          />
        </td>
        <td>
          <button type="button" onClick={this.saveCar}>
            Save
          </button>
          <button type="button" onClick={this.props.onCancelCar}>
            Cancel
          </button>
        </td>
      </tr>
    );
  }
}
