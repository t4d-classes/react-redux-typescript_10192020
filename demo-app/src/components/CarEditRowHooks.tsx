import React, { useState, ChangeEvent } from 'react';

import { Car } from '../models/cars';

export type CarEditRowProps = {
  car: Car;
  onSaveCar: (car: Car) => void;
  onCancelCar: () => void;
};

export type CarForm = {
  make: string;
  model: string;
  year: number;
  color: string;
  price: number;
};

export const CarEditRow = (props: CarEditRowProps) => {
  const [carForm, setCarForm] = useState({
    make: props.car.make,
    model: props.car.model,
    year: props.car.year,
    color: props.car.color,
    price: props.car.price,
  });

  const change = (e: ChangeEvent<HTMLInputElement>) => {
    setCarForm({
      ...carForm,
      [e.target.name]:
        e.target.type === 'number' ? Number(e.target.value) : e.target.value,
    });
  };

  const saveCar = () => {
    props.onSaveCar({ ...carForm, id: props.car.id });
  };

  return (
    <tr>
      <td>{props.car.id}</td>
      <td>
        <input
          type="text"
          id="make-input"
          name="make"
          value={carForm.make}
          onChange={change}
        />
      </td>
      <td>
        <input
          type="text"
          id="model-input"
          name="model"
          value={carForm.model}
          onChange={change}
        />
      </td>
      <td>
        <input
          type="number"
          id="year-input"
          name="year"
          value={carForm.year}
          onChange={change}
        />
      </td>
      <td>
        <input
          type="text"
          id="color-input"
          name="color"
          value={carForm.color}
          onChange={change}
        />
      </td>
      <td>
        <input
          type="number"
          id="price-input"
          name="price"
          value={carForm.price}
          onChange={change}
        />
      </td>
      <td>
        <button type="button" onClick={saveCar}>
          Save
        </button>
        <button type="button" onClick={props.onCancelCar}>
          Cancel
        </button>
      </td>
    </tr>
  );
};
