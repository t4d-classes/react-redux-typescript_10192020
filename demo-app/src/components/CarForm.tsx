import React, { Component } from 'react';
import { NewCar } from '../models/cars';

import { withForm, FormComponentProps } from '../hocs/withForm';

type CarForm = {
  make: string;
  model: string;
  year: number;
  color: string;
  price: number;
};

export type CarFormProps = {
  buttonText: string;
  onSubmitCar: (newCar: NewCar) => void;
} & FormComponentProps<CarForm>;

const getInitialState = () =>
  ({
    make: '',
    model: '',
    year: 1900,
    color: '',
    price: 0,
  } as CarForm);

export const CarForm = withForm(
  class CarForm extends Component<CarFormProps> {
    submitCar = () => {
      this.props.onSubmitCar({ ...this.props.form });
      this.props.resetForm();
    };

    render() {
      return (
        <form>
          <div>
            <label htmlFor="make-input">Make:</label>
            <input
              type="text"
              id="make-input"
              name="make"
              value={this.props.getValue('make')}
              onChange={this.props.change}
            />
          </div>
          <div>
            <label htmlFor="model-input">Model:</label>
            <input
              type="text"
              id="model-input"
              name="model"
              value={this.props.getValue('model')}
              onChange={this.props.change}
            />
          </div>
          <div>
            <label htmlFor="year-input">Year:</label>
            <input
              type="number"
              id="year-input"
              name="year"
              value={this.props.getValue('year')}
              onChange={this.props.change}
            />
          </div>
          <div>
            <label htmlFor="color-input">Color:</label>
            <input
              type="text"
              id="color-input"
              name="color"
              value={this.props.getValue('color')}
              onChange={this.props.change}
            />
          </div>
          <div>
            <label htmlFor="price-input">Price:</label>
            <input
              type="number"
              id="price-input"
              name="price"
              value={this.props.getValue('price')}
              onChange={this.props.change}
            />
          </div>
          <button type="button" onClick={this.submitCar}>
            {this.props.buttonText}
          </button>
        </form>
      );
    }
  },
  getInitialState(),
);
