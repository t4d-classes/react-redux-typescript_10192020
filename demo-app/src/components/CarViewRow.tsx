import React from "react";

import { Car } from "../models/cars";

export type CarViewRowProps = {
  car: Car;
  onEditCar: (carId: number) => void;
  onDeleteCar: (carId: number) => void;
};

export function CarViewRow(props: CarViewRowProps) {
  const editCar = () => {
    props.onEditCar(props.car.id);
  };

  const deleteCar = () => {
    props.onDeleteCar(props.car.id);
  };

  return (
    <tr>
      <td>{props.car.id}</td>
      <td>{props.car.make}</td>
      <td>{props.car.model}</td>
      <td>{props.car.year}</td>
      <td>{props.car.color}</td>
      <td>{props.car.price}</td>
      <td>
        <button type="button" onClick={editCar}>
          Edit
        </button>
        <button type="button" onClick={deleteCar}>
          Delete
        </button>
      </td>
    </tr>
  );
}
