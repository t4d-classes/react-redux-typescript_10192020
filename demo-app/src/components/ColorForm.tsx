import React, { Component, ChangeEvent } from 'react';

import { NewColor } from '../models/colors';
import { withForm, TheComponentProps } from '../hocs/withForm';

export type ColorFormProps = {
  buttonText: string;
  onSubmitColor: (color: NewColor) => void;
} & TheComponentProps;

export type ColorForm = {
  name: string;
  hexcode: string;
};

export const ColorForm = withForm(
  class ColorForm extends Component<ColorFormProps> {
    submitColor = () => {
      this.props.onSubmitColor({ ...this.props.form });
      this.props.resetForm();
    };

    render() {
      console.log('getValue: ', this.props.getValue('name'));

      return (
        <form>
          <div>
            <label htmlFor="name-input">Name:</label>
            <input
              type="text"
              id="name-input"
              name="name"
              value={this.props.getValue('name')}
              onChange={this.props.change}
            />
          </div>
          <div>
            <label htmlFor="hexcode-input">Hexcode:</label>
            <input
              type="text"
              id="hexcode-input"
              name="hexcode"
              value={this.props.getValue('hexcode')}
              onChange={this.props.change}
            />
          </div>
          <button type="button" onClick={this.submitColor}>
            {this.props.buttonText}
          </button>
        </form>
      );
    }
  },
  {
    name: '',
    hexcode: '',
  },
);
