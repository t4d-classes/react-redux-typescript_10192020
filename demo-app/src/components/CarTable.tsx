import React from 'react';

import { Car } from '../models/cars';
import { CarViewRow } from './CarViewRow';
import { CarEditRow } from './CarEditRow';

export type CarTableProps = {
  cars: Car[];
  editCarId: number;
  onEditCar: (carId: number) => void;
  onDeleteCar: (carId: number) => void;
  onSaveCar: (car: Car) => void;
  onCancelCar: () => void;
};

export function CarTable(props: CarTableProps) {
  const carTableBody = (car: Car) => {
    if (car.id === props.editCarId) {
      return (
        <CarEditRow
          key={car.id}
          car={car}
          onSaveCar={props.onSaveCar}
          onCancelCar={props.onCancelCar}
        />
      );
    } else {
      return (
        <CarViewRow
          key={car.id}
          car={car}
          onEditCar={props.onEditCar}
          onDeleteCar={props.onDeleteCar}
        />
      );
    }
  };

  return (
    <table>
      <thead>
        <tr>
          <th>Id</th>
          <th>Make</th>
          <th>Model</th>
          <th>Year</th>
          <th>Color</th>
          <th>Price</th>
          <th>Action</th>
        </tr>
      </thead>
      <tbody>{props.cars.map(carTableBody)}</tbody>
    </table>
  );
}

// {
//   props.cars.map((car) =>
//     car.id === props.editCarId ? (
//       <CarEditRow
//         key={car.id}
//         car={car}
//         onSaveCar={props.onSaveCar}
//         onCancelCar={props.onCancelCar}
//       />
//     ) : (
//       <CarViewRow
//         key={car.id}
//         car={car}
//         onEditCar={props.onEditCar}
//         onDeleteCar={props.onDeleteCar}
//       />
//     ),
//   );
// }
