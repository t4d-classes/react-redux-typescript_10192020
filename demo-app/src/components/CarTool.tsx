import React, { Component } from 'react';

import { Car, NewCar } from '../models/cars';

import { ToolHeader } from './ToolHeader';
import { CarTable } from './CarTable';
import { CarForm } from './CarForm';
import { ModalDialog } from './ModalDialog';

export type CarToolProps = {
  cars: Car[];
  editCarId: number;
  isLoading: boolean;
  onAddCar: (car: NewCar) => void;
  onSaveCar: (car: Car) => void;
  onDeleteCar: (carId: number) => void;
  onEditCar: (carId: number) => void;
  onCancelCar: () => void;
  onRefreshCars: () => void;
};

export class CarTool extends Component<CarToolProps> {
  componentDidMount() {
    this.props.onRefreshCars();
  }

  render() {
    return (
      <>
        <ToolHeader headerText="Car Tool" />
        <CarTable
          cars={this.props.cars}
          editCarId={this.props.editCarId}
          onEditCar={this.props.onEditCar}
          onDeleteCar={this.props.onDeleteCar}
          onSaveCar={this.props.onSaveCar}
          onCancelCar={this.props.onCancelCar}
        />
        <CarForm buttonText="Add Car" onSubmitCar={this.props.onAddCar} />
        {this.props.isLoading && <ModalDialog>Loading...</ModalDialog>}
      </>
    );
  }
}
