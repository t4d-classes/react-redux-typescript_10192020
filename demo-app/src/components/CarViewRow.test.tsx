import React from 'react';
import { render, RenderResult, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { Car, CarKeys } from '../models/car';
import { CarViewRow } from '../components/CarViewRow';

test('snapshot CarViewRow component', () => {
  const car: Car = {
    id: 1,
    make: 'Ford',
    model: 'Fusion Hybrid',
    year: 2020,
    color: 'red',
    price: 45000,
  };

  expect(
    renderer
      .create(
        <CarViewRow
          car={car}
          onEditCar={() => null}
          onDeleteCar={() => null}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});

describe('CarViewRow testing library', () => {
  let car: Car;
  let carKeys: CarKeys[];
  let renderResult: RenderResult;
  let editCarSpy: jest.Mock;
  let deleteCarSpy: jest.Mock;

  beforeEach(() => {
    car = {
      id: 1,
      make: 'Ford',
      model: 'Fusion Hybrid',
      year: 2020,
      color: 'red',
      price: 45000,
    };

    carKeys = Object.keys(car) as CarKeys[];

    editCarSpy = jest.fn();
    deleteCarSpy = jest.fn();
  });

  beforeEach(() => {
    renderResult = render(
      <table>
        <tbody>
          <CarViewRow
            car={car}
            onEditCar={editCarSpy}
            onDeleteCar={deleteCarSpy}
          />
        </tbody>
      </table>,
    );
  });

  test('render CarViewRow component', () => {
    const cells = renderResult.getAllByRole('cell');

    expect(cells.length).toBe(7);

    cells.slice(0, 6).map((element, index) => {
      expect(element.textContent).toBe(String(car[carKeys[index]]));
    });
  });

  test('click edit', () => {
    fireEvent.click(renderResult.getByText('Edit'));
    expect(editCarSpy).toHaveBeenCalledWith(1);
  });

  test('click delete', () => {
    fireEvent.click(renderResult.getByText('Delete'));
    expect(deleteCarSpy).toHaveBeenCalledWith(1);
  });
});
