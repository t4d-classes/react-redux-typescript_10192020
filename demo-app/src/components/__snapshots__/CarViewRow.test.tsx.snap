// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot CarViewRow component 1`] = `
<tr>
  <td>
    1
  </td>
  <td>
    Ford
  </td>
  <td>
    Fusion Hybrid
  </td>
  <td>
    2020
  </td>
  <td>
    red
  </td>
  <td>
    45000
  </td>
  <td>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Delete
    </button>
  </td>
</tr>
`;
